File: rotate_array/Rotate.java
import java.util.Arrays;

// javac Rotate.java && java Rotate
class Rotate {
	// cyclic replacement O(n)/O(1)
	static void rotate(int[] nums, int k) {
		k %= nums.length;

		int steps = 0;
		for(int startIdx=0; steps < nums.length; startIdx++){
			int currIdx = startIdx;
			int prev = nums[currIdx];
			do {
				int nextIdx = (currIdx + k) % nums.length;
				System.out.println(String.format("[%d]=>[%d]", currIdx, nextIdx));
				int tmp = nums[nextIdx];
				nums[nextIdx] = prev;
				prev = tmp;
				currIdx = nextIdx;
				steps++;
			} while(currIdx != startIdx);
		}
	}

	public static void main(String[] args) {
		// int[] nums = {1,2,3,4,5,6,7}; int k = 3; // [5,6,7,1,2,3,4]
		// int[] nums = {1,2,3,4,5}; int k = 4; // [2,3,4,5,1]
		// int[] nums = {1,2,3,4}; int k = 2; // [3,4,1,2]
		int[] nums = {1,2,3,4,5,6}; int k = 3; // [4,5,6,1,2,3]
		// int[] nums = {1,2}; int k = 3; // [2,1]
		// int[] nums = {1,2}; int k = 1; // [2,1]

		rotate(nums, k);

		System.out.println(Arrays.toString(nums));
	}
}

//       c    prev n
// [0]=>[4]=>[3]=>[2]=>[1]=>[0] (5 steps == 5 length)->stop
// startIdx = 0
// nextIdx = startIdx
//
// tmp = [nextIdx]
// [nextIdx] = prev
// prev = tmp
// nextIdx = (nextIdx + k) % size
// if nextIdx == startIdx, then finish

File: rotate_array/rotate.py
from typing import List

def rotate(nums: List[int], k: int) -> None:
	length = len(nums)
	k %= length

	startIdx = 0
	steps = 0
	while steps < length:
		prevIdx = startIdx
		prev = nums[prevIdx]

		while True:
			nextIdx = (prevIdx + k) % length
			print(f"[{prevIdx}]=>[{nextIdx}]")
			tmp = nums[nextIdx]
			nums[nextIdx] = prev
			prev = tmp
			steps += 1
			prevIdx = nextIdx
			if prevIdx == startIdx: break

		startIdx += 1


# nums = [1,2,3,4,5,6,7]; k = 3; # [5,6,7,1,2,3,4]
# nums = [1,2,3,4,5]; k = 4; # [2,3,4,5,1]
# nums = [1,2,3,4]; k = 2; # [3,4,1,2]
nums = [1,2,3,4,5,6]; k = 3; # [4,5,6,1,2,3]
# nums = [1,2]; k = 3; # [2,1]
# nums = [1,2]; k = 1; # [2,1]

rotate(nums, k)

print(nums)
